src/rust/sqs-executor/src/rusoto_helpers.rs

```diff
@@ -0,0 +1,311 @@
+    ReceiveMessageError as InnerReceiveMessageError, ReceiveMessageRequest, SendMessageRequest,
+    Sqs,
+};
+
+use rusoto_s3::PutObjectError as InnerPutObjectError;
+
+use tokio::task::{JoinError, JoinHandle};
+use tokio::time::Elapsed;
+
+use crate::errors::{CheckedError, Recoverable};
+use grapl_observe::metric_reporter::{tag, MetricReporter};
+use grapl_observe::timers::{time_fut_ms, TimedFutureExt};
+use std::io::Stdout;
+use tracing::{debug, error, warn};
+use tap::prelude::TapFallible;
+
+impl CheckedError for InnerDeleteMessageError {
+    fn error_type(&self) -> Recoverable {
+        match self {
+            InnerDeleteMessageError::InvalidIdFormat(_) => Recoverable::Persistent,
+            InnerDeleteMessageError::ReceiptHandleIsInvalid(_) => Recoverable::Persistent,
+        }
+    }
+}
+
+// PutObjectError has no variants, and conveys no information
+// about what went wrong, so we must assume a transient error
+impl CheckedError for InnerPutObjectError {
+    fn error_type(&self) -> Recoverable {
+        match self {
+            _ => Recoverable::Transient,
+        }
+    }
+}
+
+impl<E> CheckedError for RusotoError<E>
+where
+    E: CheckedError + 'static,
+{
+    /// In all cases, other than Service there's no way to inspect the error
+    /// since it's just a String, so we default to assuming it's transient, even though
+    /// that may not always be the case.
+    fn error_type(&self) -> Recoverable {
+        match self {
+            RusotoError::Service(e) => e.error_type(),
+            RusotoError::HttpDispatch(_)
+            | RusotoError::Credentials(_)
+            | RusotoError::Validation(_)
+            | RusotoError::ParseError(_)
+            | RusotoError::Unknown(_)
+            | RusotoError::Blocking => Recoverable::Transient,
+        }
+    }
+}
+
+pub async fn get_message<SqsT>(
+    queue_url: String,
+    sqs_client: SqsT,
+    metric_reporter: &mut MetricReporter<Stdout>,
+) -> Result<Vec<SqsMessage>, RusotoError<InnerReceiveMessageError>>
+where
+    SqsT: Sqs + Clone + Send + Sync + 'static,
+{
+    const WAIT_TIME_SECONDS: i64 = 20;
+    let messages = sqs_client.receive_message(ReceiveMessageRequest {
+        max_number_of_messages: Some(10),
+        queue_url,
+        visibility_timeout: Some(30),
+        wait_time_seconds: Some(WAIT_TIME_SECONDS),
+        ..Default::default()
+    });
+
+    let messages = tokio::time::timeout(
+        std::time::Duration::from_secs((WAIT_TIME_SECONDS as u64) + 1),
+        messages,
+    );
+    let (messages, ms) = time_fut_ms(messages).await;
+
+    let messages = messages
+        .expect("timeout")
+        .map(|m| m.messages.unwrap_or_else(|| vec![]));
+
+    if let Ok(ref msgs) = messages {
+        metric_reporter.histogram(
+            "sqs_executor.receive_message",
+            ms as f64,
+            &[tag("success", true), tag("empty_receive", msgs.is_empty())][..])
+            .unwrap_or_else(|e| error!(
```

Comment:
you know, this makes me wonder if metric_reporter should make all its functions .expect() - it's sort of odd to force this on the consumer